homeassistant/components/color_extractor/__init__.py

```diff
@@ -0,0 +1,171 @@
+    ATTR_RGB_COLOR,
+    DOMAIN as LIGHT_DOMAIN,
+    LIGHT_TURN_ON_SCHEMA,
+    SERVICE_TURN_ON,
+)
+from homeassistant.const import ATTR_ENTITY_ID
+from homeassistant.helpers import aiohttp_client
+import homeassistant.helpers.config_validation as cv
+
+_LOGGER = logging.getLogger(__name__)
+
+# Extend the existing light.turn_on service schema
+# with the addition of our ATTR_FILE_PATH
+FILE_SERVICE_SCHEMA = cv.make_entity_service_schema(
+    {
+        **LIGHT_TURN_ON_SCHEMA,
+        vol.Required(ATTR_ENTITY_ID): cv.string,
+        vol.Required(ATTR_FILE_PATH): cv.string,
+    }
+)
+
+URL_SERVICE_SCHEMA = cv.make_entity_service_schema(
+    {
+        **LIGHT_TURN_ON_SCHEMA,
+        vol.Required(ATTR_ENTITY_ID): cv.string,
+        vol.Required(ATTR_URL): cv.string,
+    }
+)
+
+
+def _get_file(file_path):
+    """Get a PIL acceptable input file reference.
+
+    Allows us to mock patch during testing to make BytesIO stream.
+    """
+    return file_path
+
+
+async def async_setup(hass, hass_config):
+    """Set up services for color_extractor integration."""
+
+    _LOGGER.debug("Setting up color_extractor component")
+
+    async def _async_get_color(file_handler) -> tuple:
+        """Given an image file, extract the predominant color from it."""
+        try:
+            color_thief = await hass.async_add_executor_job(ColorThief, file_handler)
+        except UnidentifiedImageError as ex:
+            _LOGGER.error("Bad image file provided, are you sure it's an image? %s", ex)
+            return
+
+        # get_color returns a SINGLE RGB value for the given image
+        color = color_thief.get_color(quality=1)
+
+        _LOGGER.debug("Extracted RGB color %s from image", color)
+
+        return color
+
+    async def _async_set_light(entity_id, color, **light_kwargs):
+        """Set the given light to our extracted RGB value."""
+        service_data = {
+            ATTR_ENTITY_ID: entity_id,
+            ATTR_RGB_COLOR: color,
+        }
+
+        # Pass the extra service call args into the light turn_on service
+        service_data.update(**light_kwargs)
+
+        _LOGGER.debug("Setting %d extra light parameters", len(light_kwargs))
+
+        _LOGGER.debug("Setting RGB %s on light %s", color, entity_id)
+
+        await hass.services.async_call(
+            LIGHT_DOMAIN, SERVICE_TURN_ON, service_data, blocking=True
+        )
+
+    async def async_predominant_color_url_service(service_call):
+        """Handle call for URL based image."""
+        service_data = dict(service_call.data)
+
+        url = service_data.pop(ATTR_URL)
+
+        if not hass.config.is_allowed_external_url(url):
+            _LOGGER.error(
+                "External URL '%s' is not allowed, please add to 'allowlist_external_urls'",
+                url,
+            )
+            return
+
+        _LOGGER.debug("Getting predominant RGB from image URL '%s'", url)
+
+        entity_id = service_data.pop(ATTR_ENTITY_ID)
```

Comment:
Why pop this? Instead, just add `ATTR_RGB_COLOR` to `service_data` and pass that on to the call service. There is no reason for `_async_set_light` to exist, just call service directly.